File: python/p0287_find_duplicate_number/solutions/bit_manipulation.py
from typing import List


class Solution:
    def findDuplicate(self, nums: List[int]) -> int:
        n = len(nums) - 1
        max_bits = n.bit_length()
        duplicate = 0

        for i in range(max_bits):
            bit_mask = 1 << i  # set the bit in position i.
            count_nums = 0
            count_range = 0
            
            # Count the number of list elements with the bit set.
            for num in nums:
                if num & bit_mask:
                    count_nums += 1
            
            # Count the number of elements in range with the bit set
            for num in range(1, n + 1):
                if num & bit_mask:
                    count_range += 1

            # If more numbers in the list have the bit set than expected, it is a duplicate.
            # Add the bit to the duplicate.
            if count_nums > count_range:
                duplicate |= bit_mask

        return duplicate

File: rust/p0287_find_duplicate_number/src/bit_manipulation.rs
pub struct Solution;

impl Solution {
    pub fn find_duplicate(nums: Vec<i32>) -> i32 {
        let n = nums.len() - 1;
        let max_bits: u32 = u32::BITS - (n as u32).leading_zeros();
        let mut duplicate: u32 = 0;

        for i in 0..max_bits {
            let bit_mask: u32 = 1 << i;
            let mut count_nums: i32 = 0;
            let mut count_range: i32 = 0;

            for &num in &nums {
                if (num as u32 & bit_mask) != 0 {
                    count_nums += 1;
                }
            }
            
            for num in 1..=n {
                if (num as u32 & bit_mask) != 0 {
                    count_range += 1;
                }
            }

            if count_nums > count_range {
                duplicate |= bit_mask;
            }
        }

        duplicate as i32
    }
}